File: fleeting_grace/simplify.py
"""Trajectory simplification using Douglas-Peucker algorithm for 3D curves."""

import numpy as np


def _point_line_distances(points: np.ndarray, start: np.ndarray, end: np.ndarray) -> np.ndarray:
    """
    Compute perpendicular distances from points to line segment (vectorized).

    Args:
        points: (N, 3) array of points
        start: (3,) line segment start
        end: (3,) line segment end

    Returns:
        (N,) array of distances
    """
    line_vec = end - start
    line_len_sq = np.dot(line_vec, line_vec)

    if line_len_sq < 1e-20:
        # Start and end are the same point
        return np.linalg.norm(points - start, axis=1)

    # Project points onto line, get parameter t
    t = np.dot(points - start, line_vec) / line_len_sq
    # Clamp to [0, 1] for segment
    t = np.clip(t, 0, 1)
    # Closest points on segment
    closest = start + np.outer(t, line_vec)
    # Distances
    return np.linalg.norm(points - closest, axis=1)


def _douglas_peucker_indices(points: np.ndarray, epsilon: float) -> list[int]:
    """
    Douglas-Peucker algorithm returning indices of points to keep.
    """
    n = len(points)
    if n <= 2:
        return list(range(n))

    # Use a stack-based approach to avoid recursion limits
    keep = np.zeros(n, dtype=bool)
    keep[0] = True
    keep[-1] = True

    stack = [(0, n - 1)]

    while stack:
        start, end = stack.pop()

        if end - start <= 1:
            continue

        # Compute distances for all points between start and end
        segment = points[start + 1 : end]
        distances = _point_line_distances(segment, points[start], points[end])

        # Find max distance
        max_idx_local = np.argmax(distances)
        max_dist = distances[max_idx_local]
        max_idx = start + 1 + max_idx_local

        if max_dist > epsilon:
            keep[max_idx] = True
            stack.append((start, max_idx))
            stack.append((max_idx, end))

    return np.where(keep)[0].tolist()


def simplify_trajectory(
    points: np.ndarray,
    epsilon: float | None = None,
) -> np.ndarray:
    """
    Simplify a 3D trajectory using Douglas-Peucker algorithm.

    Args:
        points: (N, 3) array of trajectory points
        epsilon: Maximum deviation from line (default: 0.1% of trajectory diagonal)

    Returns:
        Simplified (M, 3) array where M <= N
    """
    if len(points) <= 2:
        return points

    # Compute bounding box diagonal for default epsilon
    if epsilon is None:
        bbox_min = np.min(points, axis=0)
        bbox_max = np.max(points, axis=0)
        diagonal = np.linalg.norm(bbox_max - bbox_min)

        if diagonal < 1e-10:
            return points[:1]  # All points are the same

        epsilon = diagonal * 0.0001  # 0.01% of diagonal

    indices = _douglas_peucker_indices(points, epsilon)
    return points[indices]

File: src/simulation/simplify.ts
import type { Vec3 } from "./types";
import { sub, addScaled, lengthSq, length } from "../utils/vec3";

/**
 * Perpendicular distance from a point to a line segment.
 * Clamps the projection to [0, 1] so endpoints are handled correctly.
 */
function pointToSegmentDistance(point: Vec3, start: Vec3, end: Vec3): number {
  const seg = sub(end, start);
  const segLenSq = lengthSq(seg);

  if (segLenSq === 0) return length(sub(point, start));

  // Project point onto segment, clamped to [0, 1]
  const toPoint = sub(point, start);
  const t = Math.max(0, Math.min(1,
    (toPoint[0] * seg[0] + toPoint[1] * seg[1] + toPoint[2] * seg[2]) / segLenSq
  ));

  // Closest point on segment
  const closest = addScaled(start, seg, t);
  return length(sub(point, closest));
}

/**
 * Douglas-Peucker trajectory simplification.
 * Returns a subset of the input points that approximates the original
 * trajectory within the given epsilon tolerance.
 *
 * Uses an iterative stack-based approach to avoid recursion depth limits.
 */
export function simplifyTrajectory(points: Vec3[], epsilon?: number): Vec3[] {
  const n = points.length;
  if (n <= 2) return points.slice();

  // Default epsilon: 0.01% of trajectory bounding box diagonal
  if (epsilon === undefined) {
    epsilon = computeDefaultEpsilon(points);
  }

  const keep = new Uint8Array(n);
  keep[0] = 1;
  keep[n - 1] = 1;

  // Stack of [startIndex, endIndex] ranges to process
  const stack: [number, number][] = [[0, n - 1]];

  while (stack.length > 0) {
    const [start, end] = stack.pop()!;
    if (end - start < 2) continue;

    let maxDist = 0;
    let maxIndex = start;

    const startPt = points[start]!;
    const endPt = points[end]!;

    for (let i = start + 1; i < end; i++) {
      const dist = pointToSegmentDistance(points[i]!, startPt, endPt);
      if (dist > maxDist) {
        maxDist = dist;
        maxIndex = i;
      }
    }

    if (maxDist > epsilon) {
      keep[maxIndex] = 1;
      stack.push([start, maxIndex]);
      stack.push([maxIndex, end]);
    }
  }

  const result: Vec3[] = [];
  for (let i = 0; i < n; i++) {
    if (keep[i]) result.push(points[i]!);
  }
  return result;
}

function computeDefaultEpsilon(points: Vec3[]): number {
  let minX = Infinity, minY = Infinity, minZ = Infinity;
  let maxX = -Infinity, maxY = -Infinity, maxZ = -Infinity;

  for (const p of points) {
    if (p[0] < minX) minX = p[0];
    if (p[1] < minY) minY = p[1];
    if (p[2] < minZ) minZ = p[2];
    if (p[0] > maxX) maxX = p[0];
    if (p[1] > maxY) maxY = p[1];
    if (p[2] > maxZ) maxZ = p[2];
  }

  const dx = maxX - minX;
  const dy = maxY - minY;
  const dz = maxZ - minZ;
  const diagonal = Math.sqrt(dx * dx + dy * dy + dz * dz);

  return diagonal * 0.0001;
}
